File: scripts/generate_keyword_ledger.py
import os
import json
import re

ROOT = r"c:\Users\bunny\Downloads\00Resume\Building_Tech_Website"
ARTICLES_DIR = os.path.join(ROOT, "src", "content", "articles")
GSC_PATH = os.path.join(ROOT, "scratch", "forensic_gsc_raw.json")
LEDGER_PATH = os.path.join(ROOT, "research", "keyword_coverage_ledger.json")

def parse_frontmatter(content):
    match = re.match(r"^---\r?\n(.*?)\r?\n---", content, re.DOTALL)
    if not match:
        return {}
    fm_text = match.group(1)
    data = {}
    for line in fm_text.splitlines():
        line = line.strip()
        if ":" in line and not line.startswith("-"):
            k, v = line.split(":", 1)
            k = k.strip()
            v = v.strip().strip('"').strip("'")
            data[k] = v
    return data

files = [f for f in os.listdir(ARTICLES_DIR) if f.endswith(".mdx")]
covered = []
target_queries = set()

for f in files:
    full_path = os.path.join(ARTICLES_DIR, f)
    with open(full_path, "r", encoding="utf-8", errors="ignore") as file:
        raw = file.read()
    data = parse_frontmatter(raw)
    slug = f[:-4]
    tq = data.get("target_query", data.get("targetQuery", data.get("title", ""))).lower().strip()
    if tq:
        target_queries.add(tq)
    
    covered.append({
        "slug": slug,
        "title": data.get("title", slug),
        "targetQuery": tq,
        "category": data.get("category", "tech"),
        "draft": data.get("draft", "false").lower() == "true",
        "hasQuickAnswer": "quickAnswer:" in raw,
        "hasFaq": "faq:" in raw,
    })

uncovered = []
if os.path.exists(GSC_PATH):
    with open(GSC_PATH, "r", encoding="utf-8") as f:
        gsc_data = json.load(f)
    rows = gsc_data.get("page_query_rows", [])
    
    query_agg = {}
    for r in rows:
        q = r.get("keys", ["", ""])[1].strip().lower()
        if not q:
            continue
        if q not in query_agg:
            query_agg[q] = {
                "query": q,
                "impressions": 0,
                "clicks": 0,
                "pages": set(),
                "positions": []
            }
        item = query_agg[q]
        item["impressions"] += int(r.get("impressions", 0))
        item["clicks"] += int(r.get("clicks", 0))
        item["pages"].add(r.get("keys", [""])[0].replace("https://www.praveentechworld.com", ""))
        item["positions"].append(float(r.get("position", 0)))
    
    for q, data in query_agg.items():
        is_targeted = any(tq in q or q in tq for tq in target_queries if tq)
        avg_pos = sum(data["positions"]) / len(data["positions"]) if data["positions"] else 0
        if not is_targeted and data["impressions"] >= 15:
            uncovered.append({
                "query": q,
                "impressions": data["impressions"],
                "clicks": data["clicks"],
                "avgPosition": round(avg_pos, 1),
                "rankedPages": list(data["pages"]),
                "recommendation": "HIGH_PRIORITY_DEDICATED_ASSET" if data["impressions"] >= 100 else "EXPAND_EXISTING_SECTION"
            })

uncovered.sort(key=lambda x: x["impressions"], reverse=True)

ledger = {
    "generatedAt": "2026-09-18T23:47:00Z",
    "totalCoveredArticles": len(covered),
    "totalUncoveredOpportunities": len(uncovered),
    "coveredArticles": covered,
    "uncoveredOpportunities": uncovered,
}

os.makedirs(os.path.dirname(LEDGER_PATH), exist_ok=True)
with open(LEDGER_PATH, "w", encoding="utf-8") as f:
    json.dump(ledger, f, indent=2)

print(f"=== KEYWORD COVERAGE LEDGER GENERATED ===")
print(f"Total Covered Articles: {len(covered)}")
print(f"Uncovered GSC Queries (>=15 imp): {len(uncovered)}")
print(f"Saved to: {LEDGER_PATH}")

with open(r"c:\Users\bunny\Downloads\00Resume\Building_Tech_Website\scripts\generate_keyword_ledger.py", "w", encoding="utf-8") as f:
    with open(__file__, "r", encoding="utf-8") as src:
        f.write(src.read())

File: scripts/generate_keyword_ledger.mjs
#!/usr/bin/env node
/**
 * scripts/generate_keyword_ledger.mjs
 * Builds the comprehensive Covered vs Uncovered Keyword Ledger.
 * Maps every published article to its target query, discovers accidental queries,
 * and identifies uncovered commercial and technical keywords.
 */
import fs from 'node:fs';
import path from 'node:path';
import matter from 'gray-matter';

const ROOT = process.cwd();
const ARTICLES_DIR = path.join(ROOT, 'src', 'content', 'articles');
const GSC_PATH = path.join(ROOT, 'scratch', 'forensic_gsc_raw.json');
const LEDGER_PATH = path.join(ROOT, 'research', 'keyword_coverage_ledger.json');

const files = fs.readdirSync(ARTICLES_DIR).filter(f => f.endsWith('.mdx'));
const covered = [];
const targetQuerySet = new Set();

for (const file of files) {
  const fullPath = path.join(ARTICLES_DIR, file);
  const raw = fs.readFileSync(fullPath, 'utf-8');
  const { data } = matter(raw);
  const slug = file.replace(/\.mdx$/, '');

  const targetQuery = (data.target_query || data.targetQuery || data.title || '').toLowerCase().trim();
  if (targetQuery) {
    targetQuerySet.add(targetQuery);
  }

  covered.push({
    slug,
    title: data.title || '',
    targetQuery,
    category: data.category || 'tech',
    draft: Boolean(data.draft),
    hasQuickAnswer: Boolean(data.quickAnswer),
    hasFaq: Boolean(data.faq && data.faq.length > 0),
  });
}

// Read GSC queries
const uncovered = [];
if (fs.existsSync(GSC_PATH)) {
  const gscRaw = JSON.parse(fs.readFileSync(GSC_PATH, 'utf-8'));
  const rows = gscRaw.page_query_rows || [];

  const queryAgg = new Map();
  for (const r of rows) {
    const q = (r.keys[1] || '').trim().toLowerCase();
    if (!q) continue;
    if (!queryAgg.has(q)) {
      queryAgg.set(q, {
        query: q,
        impressions: 0,
        clicks: 0,
        pages: new Set(),
        positions: [],
      });
    }
    const item = queryAgg.get(q);
    item.impressions += Number(r.impressions) || 0;
    item.clicks += Number(r.clicks) || 0;
    item.pages.add(r.keys[0].replace('https://www.praveentechworld.com', ''));
    item.positions.push(Number(r.position) || 0);
  }

  for (const [q, data] of queryAgg.entries()) {
    const isTargeted = Array.from(targetQuerySet).some(tq => tq.includes(q) || q.includes(tq));
    const avgPos = data.positions.reduce((a, b) => a + b, 0) / data.positions.length;

    if (!isTargeted && data.impressions >= 15) {
      uncovered.push({
        query: q,
        impressions: data.impressions,
        clicks: data.clicks,
        avgPosition: Number(avgPos.toFixed(1)),
        rankedPages: Array.from(data.pages),
        recommendation: data.impressions >= 100 
          ? 'HIGH_PRIORITY_CANDIDATE_FOR_NEW_TARGET_OR_SECTION' 
          : 'SECONDARY_QUERY_TO_EXPAND_IN_EXISTING_POST',
      });
    }
  }
}

uncovered.sort((a, b) => b.impressions - a.impressions);

const ledger = {
  updatedAt: new Date().toISOString(),
  totalCoveredArticles: covered.length,
  totalUncoveredQueries: uncovered.length,
  coveredKeywords: covered,
  uncoveredOpportunities: uncovered,
};

fs.mkdirSync(path.dirname(LEDGER_PATH), { recursive: true });
fs.writeFileSync(LEDGER_PATH, JSON.stringify(ledger, null, 2), 'utf-8');

console.log(`=== KEYWORD COVERAGE LEDGER GENERATED ===`);
console.log(`Covered Articles: ${covered.length}`);
console.log(`Uncovered GSC Queries (>=15 imp): ${uncovered.length}`);
console.log(`Saved to: ${LEDGER_PATH}`);
